src/fastmcp/tools/tool.py

```diff
@@ -502,26 +502,59 @@ def _convert_to_content(
         # if the result is a list, then it could either be a list of MCP types,
         # or a "regular" list that the tool is returning, or a mix of both.
         #
-        # so we extract all the MCP types / images and convert them as individual content elements,
-        # and aggregate the rest as a single content element
-
-        mcp_types = []
-        other_content = []
+        # Group adjacent non-MCP types together while preserving order
+
+        content_items = []
+        non_mcp_batch = []
+
+        def flush_non_mcp_batch():
+            """Convert accumulated non-MCP items to a single TextContent block."""
+            if non_mcp_batch:
+                if len(non_mcp_batch) == 1:
+                    # Single item - convert directly to avoid combining when not needed
+                    content_items.extend(
+                        _convert_to_content(
+                            non_mcp_batch[0],
+                            serializer=serializer,
+                            _process_as_single_item=True,
+                        )
+                    )
+                else:
+                    # Multiple items - combine into a single text block
```

Comment:
This feels like the complexity may not be worth the benefit of compressing them all into text content... maybe its less surprising to emit a new TextContent for each one, and users who really want control should return TextContent explicitly?

Reply:
Maybe if there's any native MCP type we keep everything separate but otherwise we compress? 

It's pretty annoying for an array to get turned into 100 text content blocks but once you're returning an MCP native type maybe we assume you wanted it to be that way